### PR TITLE
Add docker support

### PR DESCRIPTION
**Pull request by clifinger**
This pull request install the Docker configuration for Bedrock project, as well as the build and extract script.

### docker-compose.yml
- Added configuration for a 3-node Bedrock cluster
- Set up networks and volumes for each node
- Defined node-specific environment variables

### Dockerfile
- Implemented a two-stage build to optimize final image size
- Used Ubuntu 22.04 as the base for the build stage
- Installed and configured apt-fast to speed up downloads
- Installed Clang 18 and necessary dependencies
- Configured ccache to improve compilation times
- Compiled the Bedrock project and ran tests
- Used phusion/baseimage:noble-1.0.0 for the final image
- Configured the Bedrock service to run with proper permissions

### build_and_extract.sh
- Added a script to build the Docker image and extract the Bedrock binary
- The script creates a temporary container, copies the binary, and places it in the output directory

## Motivation
These changes aim to improve our Docker build process, facilitate Bedrock cluster deployment, and optimize the size and performance of our Docker images.

## How to test
1. Ensure you have Docker and Docker Compose installed on your machine
2. Clone this branch and cd into `docker` folder
3. Run `docker-compose up --build` to build and start the Bedrock cluster
4. Verify that all three nodes start correctly and communicate with each other
5. Run the `build_and_extract.sh` script and check that the Bedrock binary is correctly extracted

Please feel free to provide any feedback or suggestions for improvement!

### CODE REVIEW

File: docker/libstuff/bedrock.sh
Comment (fukawi2) [inline docker/libstuff/bedrock.sh:8]:
```suggestion
  [ -z $BASH_VERSION ] && return 1
```

File: docker/libstuff/bedrock.sh
Comment (fukawi2) [inline docker/libstuff/bedrock.sh:11]:
```suggestion
  case $BASH_VERSION in
```

File: docker/libstuff/bedrock.sh
Comment (fukawi2) [inline docker/libstuff/bedrock.sh:20]: We prefer to use `[[` over `[`

```suggestion
if [[ $(bashVersionCheck) ]]; then
```

File: docker/libstuff/bedrock.sh
Comment (fukawi2) [inline docker/libstuff/bedrock.sh:29]: Our style is to keep the opening brace on the same line

```suggestion
function toLowerCase() {
```

File: docker/libstuff/bedrock.sh
Comment (fukawi2) [inline docker/libstuff/bedrock.sh:63]: Add a trailing new line please.

File: docker/libstuff/bedrock.sh
Comment (fukawi2) [inline docker/libstuff/bedrock.sh:58]:
```suggestion
for i in "${ONE_ARG_PARAMS[@]}" ; do
  if [[ -v "$i" ]]; then
```

File: docker/libstuff/bedrock.sh
Comment (fukawi2) [inline docker/libstuff/bedrock.sh:50]:
```suggestion
for i in "${NO_ARG_PARAMS[@]}" ; do
  if [[ -v "$i" ]]; then
```

**Review by fukawi2 — CHANGES_REQUESTED** [review]
Thanks for the contribution @clifinger! I've made some small suggestions to align with our style guide if that's OK.

While docker isn't something we're using or have a particular interest in right now, this seems well written, neat and self-contained. Having this available to make it easier for new users to get started seems like a good improvement too. I don't see any reason not to merge it.

What do you think @tylerkaraszewski or @coleaeason?

**Comment by clifinger** [comment]
Fixed :)

**Comment by fukawi2** [comment]
Thanks! Will wait for some others to have an opportunity to comment before I approve/merge (it's out of hours for them now)

**Comment by clifinger** [comment]
Take your time guys, btw I am the CTO of manypixels.co and I will continue to test Bedrockdb little before use it in production for our company ;)

**Comment by clifinger** [comment]
I did some tests and some changes to be sure it work:

- Added `bedrock` into `PATH` for the test
- changed `RUN make -j64` to `RUN make -j$(nproc)`

So now I am sure than I can compile and start the cluster with one command
Screenshot 2024-10-18 122247
Screenshot 2024-10-18 121704
Screenshot 2024-10-18 120902

File: docker/docker-compose.yml
Comment (coleaeason) [inline docker/docker-compose.yml:31]: should this be `8887:8887`?

File: docker/docker-compose.yml
Comment (coleaeason) [inline docker/docker-compose.yml:50]: should this be `8889:8889`?

File: docker/docker-compose.yml
Comment (coleaeason) [inline docker/docker-compose.yml:32]: should this be `9000:9000`?

File: docker/docker-compose.yml
Comment (coleaeason) [inline docker/docker-compose.yml:51]: should this be `9000:9000`?

File: docker/libstuff/bedrock.sh
Comment (coleaeason) [inline docker/libstuff/bedrock.sh:1]: just curious if there is a reason to use this bash script instead of the systemd configs in the repo and systemd to run the process?

**Comment by tylerkaraszewski** [comment]
I'm certainly not opposed to adding it.

File: docker/docker-compose.yml
Comment (clifinger) [reply docker/docker-compose.yml:31]: yes @coleaeason  it should, it mean open the port 8887 into the host and map it with the 8888 into the container, like that people can make a peer list outside of the container with :8887, 8888, 8889

File: docker/docker-compose.yml
Comment (clifinger) [reply docker/docker-compose.yml:50]: yes @coleaeason  it should, it mean open the port 8887 into the host and map it with the 8888 into the container, like that people can make a peer list outside of the container with :8887, 8888, 8889

File: docker/docker-compose.yml
Comment (clifinger) [reply docker/docker-compose.yml:32]: yes @coleaeason  it should, it mean open the port 8887 into the host and map it with the 8888 into the container, like that people can make a peer list outside of the container with :8887, 8888, 8889

File: docker/docker-compose.yml
Comment (clifinger) [reply docker/docker-compose.yml:51]: yes @coleaeason  it should, it mean open the port 8887 into the host and map it with the 8888 into the container, like that people can make a peer list outside of the container with :8887, 8888, 8889

File: docker/libstuff/bedrock.sh
Comment (clifinger) [reply docker/libstuff/bedrock.sh:1]: systemd is never include into docker because of the single process recommandation of docker. If you take a little time to read [https://github.com/phusion/baseimage-docker](https://github.com/phusion/baseimage-docker) you will understand ;)

**Comment by clifinger** [comment]
Sorry @coleaeason my answer about the ports concern the 9000+ for peers list, but it's the same meaning map the port from the host to the container. Like that you don't need to change the configuration of files inside docker when you add a node, it' s like unlimited depending of the ressource. that's the power of docker. you just add a node `node3 `with example 8881:8888 and add it to peers list inside the file docker-compose.yml and it scale out of the box.

**Comment by clifinger** [comment]
So @coleaeason you can do on the host:

```bash
nc localhost 8887
nc localhost 8888
nc localhost 8889
```
To check if it work ;) it's what I did in my tests in complements of checking the logs output

**Comment by clifinger** [comment]
2 months ago, at Manypixels, we decided to change our stack with [Dokploy](https://dokploy.com) and a dedicated server, we did it with no effort and we divided our final bill near by 10 !!!!

**Comment by clifinger** [comment]
Now if you want (free and not needed but better), you can make a docker account for Expensify and in your Github action when you merge you just add
```bash
cd docker
docker build -t expensify/bedrockdb:latest -f Dockerfile ..
docker push expensify/bedrockdb:latest
```
Like that we can change the `docker-compose.yml` from:
```yaml
bedrock-node1:
    build:
      context: ..
      dockerfile: docker/Dockerfile
```
to:
```yaml
bedrock-node1:
   image: expensify/expensify/bedrockdb:latest
```
and people will have always the latest build of the tested version ;)

File: docker/libstuff/bedrock.sh
Comment (clifinger) [reply docker/libstuff/bedrock.sh:1]: and the second argument is that we need to change a bit the configuration to handle the env. var.:

From the Readme:

## Environment Variables

The following environment variables can be used to configure Bedrock:

- `NODE_NAME`: Name of the Bedrock node
- `PRIORITY`: Priority of the node in the cluster
- `PEER_LIST`: Comma-separated list of peer nodes
- `PLUGINS`: Plugins to enable
- `CACHE_SIZE`: Size of the cache
- `WORKER_THREADS`: Number of worker threads
- `QUERY_LOG`: Query log configuration
- `MAX_JOURNAL_SIZE`: Maximum journal size
- `SYNCHRONOUS`: Synchronous mode setting

**Review by tylerkaraszewski — APPROVED** [review]

**Comment by tylerkaraszewski** [comment]
@coleaeason - I'm ok with this. I will let you resolve any outstanding issues before merging.

**Comment by clifinger** [comment]
@coleaeason where are you 😂

**Review by coleaeason — APPROVED** [review]
Thanks for the explanation! Will let @fukawi2 do a final review and merge.

File: docker/Dockerfile
Comment (fukawi2) [inline docker/Dockerfile:37]: NAB: I don't like `777` permissions, but won't block this PR since it's inside a docker container which I presume to be untrusted anyway.

File: docker/Dockerfile
Comment (fukawi2) [inline docker/Dockerfile:82]: NAB: We use `/var/lib/bedrock/` but again, doesn't really matter inside a container I guess.

**Review by fukawi2 — APPROVED** [review]
Thanks again for the contribution! ❤️

**Comment by fukawi2** [comment]
Ah damn, I missed that your commits aren't signed:

image

Are you able to rebase and sign your commits please? You can [use your SSH key](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) if you don't have a GPG key, but for compliance we do need the commits to be cryptographically signed.

File: docker/Dockerfile
Comment (clifinger) [reply docker/Dockerfile:37]: yes correct, I don't like also but for docker container it's okay

File: docker/Dockerfile
Comment (clifinger) [reply docker/Dockerfile:82]: Same no problem with that as we need to specify it

**Comment by clifinger** [comment]
> Ah damn, I missed that your commits aren't signed:
> 
> image
> 
> Are you able to rebase and sign your commits please? You can [use your SSH key](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) if you don't have a GPG key, but for compliance we do need the commits to be cryptographically signed.

Done ;)

**Comment by clifinger** [comment]
Not bad, like that my futur commits will be signed

**Comment by fukawi2** [comment]
Thanks! Merged :)

**Pull request merged**
